src/net/http/UploadFile.cpp

```diff
@@ -0,0 +1,253 @@
+  
+  size_t posted = 0;
+  size_t post_count = 0;
+
+  size_t received = 0;
+
+  OperationEnvironment &env;
+
+  Mutex mutex;
+  Cond cond;
+
+  std::exception_ptr error;
+
+  bool done = false;
+
+public:
+  UploadFileHandler(void* _buffer, size_t _max_size,
+    OperationEnvironment& _env)
+    :buffer((uint8_t*)_buffer), max_size(_max_size), env(_env) {}
+
+  void Cancel() noexcept {
+    const std::lock_guard<Mutex> lock(mutex);
+    done = true;
+    cond.notify_one();
+  }
+
+  void Wait() {
+    std::unique_lock<Mutex> lock(mutex);
+    cond.wait(lock, [this]{ return done; });
+
+    if (error)
+      std::rethrow_exception(error);
+  }
+
+  /* virtual methods from class CurlResponseHandler */
+  void OnHeaders(unsigned status,
+                 std::multimap<std::string, std::string> &&headers) override {
+    if (auto i = headers.find("content-length"); i != headers.end())
+      env.SetProgressRange(ParseUint64(i->second.c_str()));
+  }
+
+  //-----------------------------------
+  void OnData(ConstBuffer<void> data) override {
+    size_t remaining = max_size - received;
+    if (remaining == 0 || done)
+      throw Pause{};
+
+    if (data.size > remaining)
+      data.size = remaining;
+
+    memcpy(buffer + received, data.data, data.size);
+    received += data.size;
+
+    buffer[received] = 0;
+    env.SetProgressRange(received);
+  }
+
+  //-----------------------------------
+  void OnEnd() override {
+    const std::lock_guard<Mutex> lock(mutex);
+    done = true;
+    if (buffer[0] == 0)
+      std::strncpy((char*)buffer, "Upload done!", max_size - 1);
+    cond.notify_one();
+  }
+
+  //-----------------------------------
+  void OnError(std::exception_ptr e) noexcept override {
+    const std::lock_guard<Mutex> lock(mutex);
+    error = std::move(e);
+    try {
+      if (error) {
+        std::rethrow_exception(error);
+      }
+    }
+    catch (const std::exception& e) {
+      std::strncpy((char*)buffer, e.what(), max_size);
+    }
+
+    done = true;
+    cond.notify_one();    
+  }
+
+  //-----------------------------------
+  void OnUploadProgress(size_t now, size_t total) {
+    if (total > 0 && post_count == 0) {
+      post_count = total;
+      env.SetProgressRange(total);
+    }
+    if (posted < now) {
+      posted = now;
+      env.SetProgressPosition(now);
+    }
+  }
+};
+
+//-----------------------------------------------------------------------------
+static int 
+OnProgress(void* ptr, double TotalToDownload, double NowDownloaded, double TotalToUpload, double NowUploaded)
+{
+  UploadFileHandler* handler = reinterpret_cast<UploadFileHandler*> (ptr);
+  handler->OnUploadProgress(NowUploaded, TotalToUpload);
+  return 0;
+}
+
+//-----------------------------------------------------------------------------
+static struct curl_httppost* formpost = nullptr;
+static struct curl_httppost* lastptr = nullptr;
+
+//-----------------------------------------------------------------------------
```

Comment:
No ASCII art please

Reply:
I don't understand, what you mean

Reply:
You drew a horizontal line using ASCII dashes. That's ASCII art and doesn't belong there. It serves no purpose and distracts from the code.

Reply:
A pity... For me it makes more structure and a better readability of the source code. Byside the length of 80 cols show me very easy the lines they are to long. Of course I can use an other solution for this.
I will remove all this lines